src/conn/conn_tiered.c

```diff
@@ -433,9 +439,23 @@ __tier_storage_finish(WT_SESSION_IMPL *session)
 static int
 __tier_storage_copy(WT_SESSION_IMPL *session)
 {
+    WT_CONNECTION_IMPL *conn;
     WT_DECL_RET;
     WT_TIERED_WORK_UNIT *entry;
+    uint64_t ckpt_gen;
 
+    conn = S2C(session);
+    ckpt_gen = __wt_gen(session, WT_GEN_CHECKPOINT);
+    /*
+     * If the checkpoint generation has not changed, or the next checkpoint is still running we are
+     * not yet ready to copy those objects to shared storage. There is nothing to do. One global
+     * flush generation is sufficient because the next flush_tier cannot start until all flush work
+     * units from the previous call are complete. So that field is guaranteed to be the latest
+     * value.
+     */
+    if (conn->flush_gen == ckpt_gen ||
+      (conn->flush_gen < ckpt_gen && conn->txn_global.checkpoint_running))
```

Comment:
I'm wondering if there are conditions when we might get stuck here longer than we should.

1. flush_gen is set.
2. a checkpoint completes
3. another checkpoint starts (before __tiered_storage_copy is called)
4. At this point, __tier_storage_copy() is called, and the second checkpoint is running, so it returns.

Reply:
Sure, and in hindsight, checkpoint unilaterally clears the field. So it really is a boolean instead of needing the actual generation number. I think I can rename it something like `flush_checkpoint_complete` where flush_tier sets it to false and the next checkpoint sets it to true. That will also get rid of any conditions.